### PR TITLE
KAFKA-2765: Add versions to Copycat Connector and Task interfaces and log versions when instantiating connectors and tasks.

### PR DESCRIPTION
**Pull request by ewencp**

### CODE REVIEW

**Comment by ewencp** [comment]
@gwenshap You should probably review if you can. I ended up going the "duplicate" route, but showing how to avoid actually duplicating the version number by just deferring to the Connector class. I started down the path of using generics to tie the task class to the connector class, but adding them _just_ for that purpose gets messy, leaves you mostly using Tasks in their raw type format, etc.

To be honest, the more I thought about this, the more I think exposing the version info like this is going to be a stop-gap so we have something useful there, but we're eventually going to want to expose more complete metadata about the connector (e.g. embedding descriptions would be very nice, having a standard way of exposing the set of config options is probably critical to get any sort of reasonable, automatic documentation for connectors). Probably long term we'll just want to wrap version info into that more general metadata.

I started to expose the version in the REST API, but that actually requires going out to the node that's running the connector/task, whereas the current API is just getting config info out of the config topic. I think we'll eventually need to forward these requests, but we haven't figured out how to best handle that yet since only the leader knows the complete assignment of tasks... we'll eventually need to do this for status info too, but in the interest of at least getting versions into the interface so they can be used for other purposes, I didn't want to hold this up while figuring out how to manage forwarding those REST requests.

**Comment by wushujames** [comment]
I'm going to have to update https://github.com/wushujames/copycat-connector-skeleton with these changes, as well as the changes from KAFKA-2764.

Unfortunately, since I basically copy/pasted the File connector files into my skeleton, there's no "history" to show that they originally came from here, and so I'll have to apply the diffs manually to my skeleton.

**Comment by ewencp** [comment]
@wushujames shouldn't be a bit deal, these are pretty minor changes anyway. If you want to make it clear, just mention the ticket in the commit msg :)

Glad you're updating this, it is a nice resource I can point people to when they want to get started writing a connector! We're trying to round out any final changes to APIs, but I think this was the last of it that should affect the skeleton project.

**Comment by wushujames** [comment]
Yeah, I'm not worried. It's just another thing to do. And yeah, once the API is stable, I'll only have to update it if you change the API, and that would be infrequent (on a Kafka release boundary, most likely).

Perforce allows forking at the file level, and maintains history. So I can fork tree1/fileA to tree2/fileA, and then merge over future changes of tree1/fileA into tree2/fileA and it'll keep track of things. Kinda nice. Can't do that in git or mercurial, or at least I haven't figured it out.

I personally found the skeleton useful to create my MySQL connector! :)

**Comment by gwenshap** [comment]
Yeah, agree that we'll need more complete metadata for connectors. Especially for the much anticipated connector discovery service ;)

File: copycat/file/src/main/java/org/apache/kafka/copycat/file/FileStreamSinkConnector.java [hunk 1]
```diff
@@ -35,6 +36,11 @@
     private String filename;
 
     @Override
+    public String version() {
+        return AppInfoParser.getVersion();
```
Comment (gwenshap) [inline copycat/file/src/main/java/org/apache/kafka/copycat/file/FileStreamSinkConnector.java:40]: awesome

File: copycat/file/src/main/java/org/apache/kafka/copycat/file/FileStreamSinkTask.java [hunk 2]
```diff
@@ -50,6 +50,11 @@ public FileStreamSinkTask(PrintStream outputStream) {
     }
 
     @Override
+    public String version() {
+        return new FileStreamSinkConnector().version();
```
Comment (gwenshap) [inline copycat/file/src/main/java/org/apache/kafka/copycat/file/FileStreamSinkTask.java:54]: also awesome :+1:

**Comment by gwenshap** [comment]
LGTM.

**Pull request closed without merging**